### PR TITLE
Fix for error with gcc8

### PR DESCRIPTION
**Pull request by npanpaliya**
## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/open-ce/open-ce/blob/main/CONTRIBUTING.md)?
- [ ] Did you update any affected [documentation](https://github.com/open-ce/open-ce/blob/main/doc/)?
- [ ] Did you write any tests to validate this change?

## Description

Fixes part of the build error with gcc 8. There is one more error which is trickier and as per gcc's documentation, it is fixed in gcc 9.1
```
/mnt/pai/home/npanpa23/anaconda3/conda-bld/dali_1627563466949/work/dali/operators/image/remap/warp.h:319:20: internal compiler error: in tsubst_copy, at cp/pt.c:15478
               auto param_provider = This().template CreateParamProvider<spatial_ndim, BorderType>();
                    ^~~~~~~~~~~~~~
/mnt/pai/home/npanpa23/anaconda3/conda-bld/dali_1627563466949/work/third_party/boost/preprocessor/include/boost/preprocessor/facilities/identity.hpp:23:34: note: in definition of macro 'BOOST_PP_IDENTITY'
 # define BOOST_PP_IDENTITY(item) item BOOST_PP_EMPTY
                                  ^~~~
/mnt/pai/home/npanpa23/anaconda3/conda-bld/dali_1627563466949/work/third_party/boost/preprocessor/include/boost/preprocessor/tuple/rem.hpp:123:49: note: in expansion of macro 'BOOST_PP_REM'
 #     define BOOST_PP_TUPLE_REM_CTOR_O_1(tuple) BOOST_PP_REM tuple
                                                 ^~~~~~~~~~~~
/mnt/pai/home/npanpa23/anaconda3/conda-bld/dali_1627563466949/work/third_party/boost/preprocessor/include/boost/preprocessor/cat.hpp:29:34: note: in expansion of macro 'BOOST_PP_TUPLE_REM_CTOR_O_1'
 #    define BOOST_PP_CAT_I(a, b) a ## b
```

## Review process to land 

1. All tests and other checks must succeed.
2. At least one [maintainer](https://github.com/open-ce/open-ce/blob/main/MAINTAINERS.md) must review and approve.
3. If any  [maintainer](https://github.com/open-ce/open-ce/blob/main/MAINTAINERS.md) requests changes, they must be addressed.

### CODE REVIEW

File: recipe/0102-Fix-error-using-gcc-8.patch [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+diff --git a/dali/operators/audio/mfcc/mfcc.h b/dali/operators/audio/mfcc/mfcc.h
+index c6d4d06e..eb55f03c 100644
+--- a/dali/operators/audio/mfcc/mfcc.h
++++ b/dali/operators/audio/mfcc/mfcc.h
+@@ -88,22 +88,22 @@ class MFCC : public Operator<Backend> {
+   void GetArguments(const workspace_t<Backend> &ws) {
+     auto nsamples = ws.template InputRef<Backend>(0).shape().size();
+     DctArgs arg;
+-    arg.ndct = spec_.template GetArgument<int>("n_mfcc");
```
Comment (JanuszL) [inline recipe/0102-Fix-error-using-gcc-8.patch:18]: Why not make a PR directly to DALI?

Comment (jayfurmanek) [reply recipe/0102-Fix-error-using-gcc-8.patch:18]: This is just a back port of: https://github.com/NVIDIA/DALI/pull/1320

Comment (jayfurmanek) [reply recipe/0102-Fix-error-using-gcc-8.patch:18]: Well, relooking It does look like we have that fix in a few other places.
Gcc8 also seems to choke on this block, which may need the same:
https://github.com/NVIDIA/DALI/blob/v1.3.0/dali/operators/image/remap/warp.h#L317-L321

Comment (jayfurmanek) [reply recipe/0102-Fix-error-using-gcc-8.patch:18]: We'll PR upsteam if this all works.

Comment (npanpaliya) [reply recipe/0102-Fix-error-using-gcc-8.patch:18]: Sure, I'll raise the PR DALI too.

Comment (JanuszL) [reply recipe/0102-Fix-error-using-gcc-8.patch:18]: The change is similar to NVIDIA/DALI#1320, but is about mfcc.h.
If this works we would be happy to include it in the mainline.

**Comment by jayfurmanek** [comment]
GCC 8 (or newer) needed because of this patch (which is also upstream TF now):
https://github.com/open-ce/tensorflow-feedstock/pull/55

**Comment by JanuszL** [comment]
Hi @npanpaliya,
> @JanuszL - Do you have any pointers on the above mentioned error seen with building DALI with gcc 8?

We are able to build DALI with gcc 7.x and 9.x. I haven't tested with 8.x to be honest.
Which particular error do you have in mind from the one mentioned in this thread?

**Comment by npanpaliya** [comment]
@JanuszL - 
I tried with gcc 9 too and got failure with that too on ppc64le.

The two errors I see are as follows -
```1. 
/mnt/pai/home/npanpa23/anaconda3/conda-bld/dali_1633604179151/work/dali/operators/image/remap/warp.h:319:20: internal compiler error: in tsubst_copy, at cp/pt.c:15478
               auto param_provider = This().template CreateParamProvider<spatial_ndim, BorderType>();
```
```2. 
/mnt/pai/home/npanpa23/anaconda3/conda-bld/dali_1633604179151/work/dali/kernels/slice/slice_cpu.h:317:20: error: 'void* memcpy(void*, const void*, size_t)' copying an object of non-trivial type 'struct dali::float16' from an array of 'const double' [-Werror=class-memaccess]
         std::memcpy(out_data + b_start, in_data + b_start, b_nbytes);
         ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mzient** [comment]
Hello, @npanpaliya !
The 1st error i s clearly a GCC bug. See the minimal repro:
https://godbolt.org/z/8svM1x1z7
If you switch to any other version of GCC (or clang, or NVCC), it compiles.
In principle, it's fixable and if you think it's worthwhile to have DALI compile with GCC 8, you can always file a PR with necessar changes.

However, if that's related to TF, there's no need to match the compiler with which entire DALI is built to Tensorflow - only the DALI TF plugin must match.

**Comment by npanpaliya** [comment]
Hi @mzient.
Thank you for your reply. I do agree that the first one is a gcc bug and have several links/bugs that proves it. 
I also agree that DALI can be built with any gcc irrespective of the GCC version used to build TF. I too tried gcc 7 and DALI built fine but DALI TF plugin fails saying `vec_neg` not declared in the scope. (And this error about `vec_neg` is clearly said to be fixed in gcc 8.1 onwards.) But since both DALI and DALI TF plugin are built in the same recipe, I can not change GCC in between the builds.  
So, the options I've is to either use gcc>=8 or separate out the recipes for DALI and DALI TF plugin.
Currently, I'm trying with gcc 9.4 and if it doesn't work, I'll try separating the recipes for both.

**Comment by npanpaliya** [comment]
Finally I could build DALI and DALI TF plugin using gcc 9.4. Thank you @JanuszL and @mzient.

**Pull request closed without merging**